File: python/src/agent_framework/core/channel_adapter.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Awaitable, Callable


@dataclass
class NormalizedMessage:
    channel: str
    peer: str
    text: str
    metadata: dict | None = None


class InMemoryChannelAdapter:
    def __init__(self, channel_id: str) -> None:
        self.id = channel_id
        self._handler: Callable[[NormalizedMessage], Awaitable[None]] | None = None

    async def start(self, on_message: Callable[[NormalizedMessage], Awaitable[None]]) -> None:
        self._handler = on_message

    async def stop(self) -> None:
        self._handler = None

    async def send(self, peer: str, text: str) -> None:
        del peer, text

    async def inject(self, peer: str, text: str) -> None:
        if self._handler is None:
            raise RuntimeError(f"Channel {self.id} is not started")
        await self._handler(NormalizedMessage(channel=self.id, peer=peer, text=text))

File: typescript/packages/core/src/channel-adapter.ts
export interface NormalizedMessage {
  channel: string;
  peer: string;
  text: string;
  metadata?: Record<string, unknown>;
}

export interface ChannelAdapter {
  readonly id: string;
  start(onMessage: (message: NormalizedMessage) => Promise<void>): Promise<void>;
  stop(): Promise<void>;
  send(peer: string, text: string): Promise<void>;
}

export class InMemoryChannelAdapter implements ChannelAdapter {
  readonly id: string;
  private handler: ((message: NormalizedMessage) => Promise<void>) | null = null;

  constructor(id: string) {
    this.id = id;
  }

  async start(onMessage: (message: NormalizedMessage) => Promise<void>): Promise<void> {
    this.handler = onMessage;
  }

  async stop(): Promise<void> {
    this.handler = null;
  }

  async send(peer: string, text: string): Promise<void> {
    void peer;
    void text;
  }

  async inject(peer: string, text: string): Promise<void> {
    if (!this.handler) {
      throw new Error(`Channel ${this.id} is not started`);
    }
    await this.handler({ channel: this.id, peer, text });
  }
}
